samcli/lib/telemetry/metric.py

```diff
@@ -138,15 +242,21 @@ def wrapped(*args, **kwargs):
 
         try:
             ctx = Context.get_current_context()
-            metric.add_data("awsProfileProvided", bool(ctx.profile))
-            metric.add_data("debugFlagProvided", bool(ctx.debug))
-            metric.add_data("region", ctx.region or "")
-            metric.add_data("commandName", ctx.command_path)  # Full command path. ex: sam local start-api
-            # Metric about command's execution characteristics
-            metric.add_data("duration", duration_fn())
-            metric.add_data("exitReason", exit_reason)
-            metric.add_data("exitCode", exit_code)
-            telemetry.emit(metric)
+            track_metric(
+                MetricName.commandRun,
+                top_level_attrs={
+                    "awsProfileProvided": bool(ctx.profile),
+                    "debugFlagProvided": bool(ctx.debug),
+                    "region": ctx.region or "",
+                    "commandName": ctx.command_path,
+                    "duration": duration_fn(),
+                    "exitReason": exit_reason,
+                    "exitCode": exit_code,
+                },
+                # NOTE: setting overwrite_existing_top_level_attrs to True since this is run after the commd logic runs
+                # the top_level_attrs tracked here takes precedence
+                overwrite_existing_top_level_attrs=True,
```

Comment:
I still dont fully understand this, is there a order in which the track_metric gets called, where the overwriting is safe?

Reply:
The `track_metric` here is called after all command behaviors complete. It shouldn't happen but we could mistakenly write some of the attributes (e.g. `region`) in previous `track_metric` calls throughout the session.

Thinking out loud - maybe we should have two `track_metric` methods instead:
* `track_metric` - safe and doesn't come with the capabilitiy of overwriting existing attributes. This should be used as a generic method used throughout the session
* `_track_metric_unsafe` - unsafe in a sense that it can overwrite existing attributes. It should only be used here, where overwriting is necessary